Serial controller

lulf opened the pull request:
It works with my zephyr-over-uart   with microbit <-> nrf52dk (zephyr). 

* Not entirely sure about the safety for sync commands. If the caller drops the future, the slot should be 'released' but might need better safeguarding?
* Not using command handle, for now only 1 request per opcode.

alexmoon commented on src/serial.rs:
```diff
@@ -0,0 +1,325 @@
+{
+    fn exec(&self, cmd: &C) -> impl Future<Output = Result<C::Return, CmdError<Self::Error>>> {
+        async {
+            // TODO: Something more appropriately sized matching the expected return lenght
```
We could add a `const RETURN_PARAMS_LEN_MAX: usize` to the `SyncCmd` trait. Easily calculated for fixed-sized return parameters. For variable length return parameters the spec specifies a max of 252 bytes.

lulf replied:
Unforunately I'm running into `constant expression depends on a generic parameter` error for that. I have a feeling it should be possible somehow.

alexmoon replied:
Yeah, I've run into that error a few times. I've gone ahead and implemented a solution in 2a56894. Let me know your thoughts.

lulf replied:
Ah, I was on that path actually but didn't work as I wasn't using a separate trait + assoc constant for the buf. I think it looks great!

alexmoon commented on src/serial.rs:
```diff
@@ -0,0 +1,325 @@
+        async {
+            loop {
+                {
+                    // Safety: we will not hold references across loop iterations.
+                    let buf = unsafe { core::slice::from_raw_parts_mut(buf.as_mut_ptr(), buf.len()) };
```
That's annoying, but I don't know a good way around it.

alexmoon commented on src/serial.rs:
```diff
@@ -0,0 +1,325 @@
+impl<M, R, W, const SLOTS: usize> Controller for SerialController<M, R, W, SLOTS>
+where
+    M: RawMutex,
+    R: embedded_io_async::Read,
+    W: embedded_io_async::Write,
+{
+    type Error = embedded_io::ErrorKind;
```
If you constrain `R` and `W` to have the same error type, you could use that common error type here.

lulf replied:
Fixed

alexmoon reviewed (commented):
Looking really good! We may want to split this into a separate, associated, crate at some point. The main thing I think is missing is the HCI command flow control (see the core spec Vol 4, Part E, Section 4.4 for details).

I'm thinking we'd add a semaphore to the `ControllerState` (such as [this](https://docs.rs/futures-intrusive/latest/futures_intrusive/sync/struct.GenericSemaphore.html) from futures-intrusive). When acquiring a slot you'd first acquire a permit. When receiving CommandComplete or CommandStatus you would reset the number of permits to the number to the `num_hci_cmd_pkts` parameter.

The other thing that needs special handling is the `Reset` command. It should consume all available permits (so no other commands can be sent), and when it completes all other outstanding commands should be signaled with an error.

lulf commented:
@alexmoon Thanks for the review! I've split the SerialController into a HciController, a HciDriver trait and a SerialHciDriver implementation. This is to more easily support packet-oriented HCI interfaces such as Raspberry Pico W. 

I'll address your other comments shortly.

lulf commented:
@alexmoon I've addressed most of the points I believe. Agreed that maybe this belongs elsewhere, I'm wondering now if maybe the driver and serial modules is a better to be moved to trouble. I'd like to do HCI data flow control as well, but this I think needs some more integration with trouble anyway.

alexmoon commented:
> @alexmoon Thanks for the review! I've split the SerialController into a HciController, a HciDriver trait and a SerialHciDriver implementation. This is to more easily support packet-oriented HCI interfaces such as Raspberry Pico W.

I like the driver trait. That looks good to me.

> @alexmoon I've addressed most of the points I believe. Agreed that maybe this belongs elsewhere, I'm wondering now if maybe the driver and serial modules is a better to be moved to trouble. I'd like to do HCI data flow control as well, but this I think needs some more integration with trouble anyway.

I think the controller implementations should be independent of the host, just in case anyone else is crazy enough to build a separate host implementation. At the moment I think we can keep it in bt-hci, behind a feature flag.

lulf commented:
> > @alexmoon Thanks for the review! I've split the SerialController into a HciController, a HciDriver trait and a SerialHciDriver implementation. This is to more easily support packet-oriented HCI interfaces such as Raspberry Pico W.
> 
> I like the driver trait. That looks good to me.
> 
> > @alexmoon I've addressed most of the points I believe. Agreed that maybe this belongs elsewhere, I'm wondering now if maybe the driver and serial modules is a better to be moved to trouble. I'd like to do HCI data flow control as well, but this I think needs some more integration with trouble anyway.
> 
> I think the controller implementations should be independent of the host, just in case anyone else is crazy enough to build a separate host implementation. At the moment I think we can keep it in bt-hci, behind a feature flag.

Sounds good to me, it's more easy to maintain this way as well :+1: All good to merge then? I have a pending PR for nrf-sdc for these API changes as well.

alexmoon commented:
I've made a few more changes. The biggest change is to `HciDriver`  (now named `Transport`, see below) to operate on `ControllerToHostPacket` and `HostToControllerPacket` rather than byte arrays. This allows some transports (like the `SerialTransport`) to avoid having to serialize the packet to an extra buffer before writing it out or having to partially deserialize an incoming packet first before reading it into a buffer just to be deserialized again. Also, some transports do not use the packet type indicator byte so this avoids having to serialize that byte into a buffer just for it to be ignored.

Otherwise I just moved things around and changed some names to fit a bit better:
- I renamed `HciDriver` to `Transport` (to match the terminology from the Bluetooth spec) and moved it into a `transport` module.
- Similarly, `SerialHciDriver` becomes `SerialTransport`, also in the `transport` module
- I moved the `Controller` traits into a `controller` module
- I renamed `HciController` to `ExternalController` and moved it into the `controller` module

With these changes, I think we're ready to merge. Let me know if any of these changes cause you problems.

lulf commented:
Looks great to me!
